site/app/Panbench/Shake/Lang/Agda.hs

```diff
@@ -0,0 +1,118 @@
+  , needAgda
+  -- $shakeAgdaCommands
+  , agdaCheckDefaultArgs
+  , agdaDoctor
+  -- $shakeAgdaRules
+  , agdaRules
+  ) where
+
+import Data.Char
+
+import Development.Shake
+import Development.Shake.Classes
+
+import GHC.Generics
+
+import Panbench.Shake.AllCores
+import Panbench.Shake.File
+import Panbench.Shake.Git
+import Panbench.Shake.Store
+
+import System.Directory qualified as Dir
+import System.FilePath
+
+-- * Agda Installation
+--
+-- $shakeAgdaInstall
+
+-- | Query for installing a version of @agda@.
+data AgdaQ = AgdaQ
+  { agdaInstallRev :: String
+  -- ^ Revision of Agda to install.
+  , agdaInstallFlags :: [String]
+  -- ^ Compile flags used to build Agda.
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving anyclass (Hashable, Binary, NFData)
+
+-- | Default flags to use for Agda installation.
+--
+-- We use the following defaults:
+-- * @-foptimise-heavily@ is the default for release builds of @agda@.
+-- * @--disable-documentation@, as we don't need haddocks.
+-- * @--disable-tests@, as we don't need to run the test suite.
+-- * @--disable-profiling@, as we aren't creating profiled builds.
+defaultAgdaInstallFlags :: [String]
+defaultAgdaInstallFlags =
+  [ "-foptimise-heavily"
+  , "--disable-documentation"
+  , "--disable-tests"
+  , "--disable-profiling"
+  ]
+
+-- | Oracle for installing a version of Agda.
+--
+-- The oracle returns the absolute path to the produced @agda@ binary.
+agdaInstall :: AgdaQ -> FilePath -> Action ()
+agdaInstall AgdaQ{..} storeDir = do
+  let repoDir = "_build/repos/agda"
+  let workDir = replaceDirectory storeDir "_build/repos"
+  needGitWorktree $ GitWorktreeQ
+    { gitWorktreeUpstream = "https://github.com/agda/agda.git"
+    , gitWorktreeRepo = repoDir
+    , gitWorktreeDir = workDir
+    , gitWorktreeRev = agdaInstallRev
+    }
+  -- [TODO: Reed M, 14/07/2025] We could be more reproducible by allowing the
+  -- user to specify a cabal lockfile.
+  --
+  -- Note that this also uses the system GHC: we could make this more configurable by
+  -- calling out to @ghcup@, but let's just get things working for now
+  withAllCores \nCores -> do
+    command_ [Cwd workDir] "cabal" (["build", "agda", "--project-dir=.", "--jobs=" ++ show nCores] ++ agdaInstallFlags)
+  Stdout listBinOut <- command [Cwd workDir] "cabal" (["list-bin", "agda", "--project-dir=."] ++ agdaInstallFlags)
+  let outDir = takeDirectory $ takeWhile (not . isSpace) listBinOut
+  copyDirectoryRecursive outDir storeDir
+
+-- | Require that a particular version of @agda@ is installed,
+-- and return the absolute path pointing to the executable.
+needAgda :: AgdaQ -> Action FilePath
+needAgda q = do
+  (store, _) <- askStoreOracle q
+  path <- liftIO $ Dir.makeAbsolute (store </> "agda")
+  pure path
+
+-- * Running Agda
+--
+-- $shakeAgdaCommands
+
+-- | Default arguments for @agda@ to check a file.
+agdaCheckDefaultArgs :: FilePath -> [String]
+agdaCheckDefaultArgs file = ["+RTS", "-M3.0G", "-RTS", file]
```

Comment:
3Gigs seems small, no? And I guess all systems have a similar command?

Reply:
Or is this just for testing that Agda works rather than for all tests?

Reply:
These flags were taken from the go CLI, but yes, 3 gigs is on the low end!